kv/kvnemesis: TestKVNemesisSingleNode failed

kv/kvnemesis.TestKVNemesisSingleNode [failed](https://tanzanite.cluster.engflow.com/invocations/default/da29002a-1e41-4562-9347-1324956473a8?testReportRun=7&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on master @ [460707194644edbf4c6258d09b46d9877c8eefa6](https://github.com/cockroachdb/cockroach/commits/460707194644edbf4c6258d09b46d9877c8eefa6):


```
=== RUN   TestKVNemesisSingleNode
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestKVNemesisSingleNode3458089655
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:328: seed: 6297214651202231189
    kvnemesis_test.go:235: kvnemesis logging to /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_7_of_30/test.outputs/kvnemesis3065144168
    kvnemesis.go:178: error applying x.AdminSplit(ctx, tk(17558946769533644915), hlc.MaxTimestamp) // failed to re-compute MVCCStats pre split: descriptor mismatch; range likely merged: failed to re-compute MVCCStats pre split: descriptor mismatch; range likely merged
    kvnemesis.go:213: failures(verbose): /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_7_of_30/test.outputs/kvnemesis3065144168/failures
        repro steps: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_7_of_30/test.outputs/kvnemesis3065144168/repro.go
        rangefeed KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_7_of_30/test.outputs/kvnemesis3065144168/kvs-rangefeed.txt
        scan KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_7_of_30/test.outputs/kvnemesis3065144168/kvs-scan.txt
    kvnemesis_test.go:356: [0] proposed: 815
    kvnemesis_test.go:358: [0] reproposed unchanged: 66
    kvnemesis_test.go:360: [0] reproposed with new LAI: 7
    kvnemesis_test.go:365: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:365
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:275
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisSingleNode
        	Messages:   	kvnemesis detected failures
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestKVNemesisSingleNode3458089655
--- FAIL: TestKVNemesisSingleNode (6.97s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37160

--- a/pkg/kv/kvserver/batcheval/cmd_recompute_stats.go
+++ b/pkg/kv/kvserver/batcheval/cmd_recompute_stats.go
@@ -61,6 +61,11 @@ func declareKeysRecomputeStats(
 	return nil
 }
 
+// RecomputeStatsMismatchError indicates that the start key provided in the
+// request arguments doesn't match the start key of the range descriptor. This
+// can happen when a concurrent merge subsumed this range into another one.
+var RecomputeStatsMismatchError = errors.New("descriptor mismatch; range likely merged")
+
 // RecomputeStats recomputes the MVCCStats stored for this range and adjust them accordingly,
 // returning the MVCCStats delta obtained in the process.
 func RecomputeStats(
@@ -69,7 +74,7 @@ func RecomputeStats(
 	desc := cArgs.EvalCtx.Desc()
 	args := cArgs.Args.(*kvpb.RecomputeStatsRequest)
 	if !desc.StartKey.AsRawKey().Equal(args.Key) {
-		return result.Result{}, errors.New("descriptor mismatch; range likely merged")
+		return result.Result{}, RecomputeStatsMismatchError
 	}
 	dryRun := args.DryRun
 

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -1185,6 +1185,93 @@ func TestStoreRangeSplitWithTracing(t *testing.T) {
 	require.NotRegexp(t, traceRegexp, stringifiedSpans)
 }
 
+// TestStoreRangeSplitWithMismatchedDesc reproduces a split failure when the
+// request to re-compute stats fails due to a range descriptor mismatch.
+// It does so by splitting a range and concurrently subsuming that range via a
+// range merge. The split fails and returns an error to the client.
+func TestStoreRangeSplitWithMismatchedDesc(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	splitBlocked := make(chan struct{})
+	doneBlocking := atomicValue[bool]{}
+	doneBlocking.set(false)
+	filter := func(ctx context.Context, request *kvpb.BatchRequest) *kvpb.Error {
+		if req, ok := request.GetArg(kvpb.RecomputeStats); ok {
+			rs := req.(*kvpb.RecomputeStatsRequest)
+			// Block the split right before evaluating the RecomputeStats request.
+			// Only do so the first time; the split may get retried or there could be
+			// other RecomputeStatsRequests, and we don't want to block them.
+			if rs.Key.Equal(roachpb.Key("b")) && !doneBlocking.get() {
+				// Signal that the split is blocked.
+				splitBlocked <- struct{}{}
+				// Wait for split to be unblocked.
+				<-splitBlocked
+				doneBlocking.set(true)
+			}
+		}
+		return nil
+	}
+
+	ctx := context.Background()
+	settings := cluster.MakeClusterSettings()
+	kvserver.EnableEstimatedMVCCStatsInSplit.Override(ctx, &settings.SV, true)
+	kvserver.EnableMVCCStatsRecomputationInSplit.Override(ctx, &settings.SV, true)
+	s := serverutils.StartServerOnly(t, base.TestServerArgs{
+		Knobs: base.TestingKnobs{
+			Store: &kvserver.StoreTestingKnobs{
+				DisableMergeQueue:    true,
+				DisableSplitQueue:    true,
+				TestingRequestFilter: filter,
+			},
+		},
+		Settings: settings,
+	})
+
+	defer s.Stopper().Stop(ctx)
+	store, err := s.GetStores().(*kvserver.Stores).GetStore(s.GetFirstStoreID())
+	require.NoError(t, err)
+
+	// Write some initial data.
+	_, pErr := kv.SendWrapped(ctx, store.TestSender(), putArgs([]byte("a"), []byte("foo")))
+	require.NoError(t, pErr.GoError())
+	_, pErr = kv.SendWrapped(ctx, store.TestSender(), putArgs([]byte("b"), []byte("bar")))
+	require.NoError(t, pErr.GoError())
+	_, pErr = kv.SendWrapped(ctx, store.TestSender(), putArgs([]byte("c"), []byte("foo")))
+	require.NoError(t, pErr.GoError())
+	_, pErr = kv.SendWrapped(ctx, store.TestSender(), putArgs([]byte("d"), []byte("bar")))
+	require.NoError(t, pErr.GoError())
+
+	// Split the range at "b".
+	_, pErr = kv.SendWrapped(ctx, store.TestSender(), adminSplitArgs(roachpb.Key("b")))
+	require.NoError(t, pErr.GoError())
+
+	// Split the new RHS at "c". This split will be blocked by the above filter
+	// because the range we're splitting starts at key "b".
+	g := ctxgroup.WithContext(ctx)
+	g.GoCtx(func(ctx context.Context) error {
+		// Use AdminSplit with a non-test sender here to reproduce exactly what will
+		// be returned to the client.
+		return store.DB().AdminSplit(ctx, roachpb.Key("c"), hlc.MaxTimestamp)
+	})
+
+	// Wait until split is underway.
+	<-splitBlocked
+
+	// Merge the LHS and RHS back into one range ("a" - "d").
+	_, pErr = kv.SendWrapped(ctx, store.TestSender(), adminMergeArgs(roachpb.Key("a")))
+	require.NoError(t, pErr.GoError())
+
+	// Unblock the split.
+	splitBlocked <- struct{}{}
+
+	// Wait for the split to complete. The RecomputeStats requests finds a
+	// mismatch between the start key of the request ("b") and the start key of
+	// the merged range ("a").
+	require.Regexp(t, g.Wait(),
+		"failed to re-compute MVCCStats pre split: descriptor mismatch; range likely merged")
+}
+
 // RaftMessageHandlerInterceptor wraps a storage.IncomingRaftMessageHandler. It
 // delegates all methods to the underlying storage.IncomingRaftMessageHandler,
 // except that HandleSnapshot calls receiveSnapshotFilter with the snapshot

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -1185,10 +1185,11 @@ func TestStoreRangeSplitWithTracing(t *testing.T) {
 	require.NotRegexp(t, traceRegexp, stringifiedSpans)
 }
 
-// TestStoreRangeSplitWithMismatchedDesc reproduces a split failure when the
-// request to re-compute stats fails due to a range descriptor mismatch.
-// It does so by splitting a range and concurrently subsuming that range via a
-// range merge. The split fails and returns an error to the client.
+// TestStoreRangeSplitWithMismatchedDesc ensures that if a RecomputeRequest
+// during a split fails due to a range descriptor mismatch, the split doesn't
+// return an error to the client.
+// The test works by splitting a range and concurrently subsuming that range via
+// a range merge. The split doesn't return an error to the client.
 func TestStoreRangeSplitWithMismatchedDesc(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
@@ -1265,11 +1266,11 @@ func TestStoreRangeSplitWithMismatchedDesc(t *testing.T) {
 	// Unblock the split.
 	splitBlocked <- struct{}{}
 
-	// Wait for the split to complete. The RecomputeStats requests finds a
-	// mismatch between the start key of the request ("b") and the start key of
-	// the merged range ("a").
-	require.Regexp(t, g.Wait(),
-		"failed to re-compute MVCCStats pre split: descriptor mismatch; range likely merged")
+	// Wait for the split to complete and ensure no error is returned to the client.
+	// The split will hit a benign error because the range descriptor changed, and
+	// will be retried. The second time around, the split is a no-op because the
+	// range "b" - "d" was subsumed.
+	require.Nil(t, g.Wait())
 }
 
 // RaftMessageHandlerInterceptor wraps a storage.IncomingRaftMessageHandler. It

--- a/pkg/kv/kvserver/replica_command.go
+++ b/pkg/kv/kvserver/replica_command.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/allocatorimpl"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/plan"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/allocator/storepool"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/benignerror"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
@@ -516,7 +517,17 @@ func (r *Replica) adminSplitWithDescriptor(
 			var ba kv.Batch
 			ba.AddRawRequest(&req)
 			if err = r.store.DB().Run(ctx, &ba); err != nil {
-				return reply, errors.Wrapf(err, "failed to re-compute MVCCStats pre split")
+				// An error here is most likely because a descriptor mismatch was
+				// detected in cmd_recompute_stats. Unfortunately, that error is
+				// different from the descriptor-changed errors handled here in
+				// replica_command; the latter present as ConditionFailedError. We log
+				// the error here, and let the code below handle the descriptor-changed
+				// error coming from splitTxnAttempt.
+				if errors.Is(err, batcheval.RecomputeStatsMismatchError) {
+					log.Warningf(ctx, "failed to re-compute MVCCStats pre split: %v", err)
+				} else {
+					return reply, errors.Wrapf(err, "failed to re-compute MVCCStats pre split")
+				}
 			}
 		}
 

